Update binary format documentation & Python sample code (#3163)
XRPLF/xrpl-dev-portal: docs/references/protocol/binary-format.md

### Amount Fields
[Amount]: #amount-fields

The _Amount_ type (also called "STAmount") is a special field type that represents an amount of currency or asset. This type consists of three sub-types, which are XRP, fungible tokens, and multi-purpose tokens (MPTs):

- **XRP**

    XRP is serialized as a 64-bit unsigned integer (big-endian order), except that the most significant bit is always `0`, the second-most-significant bit is `1` to indicate that it is positive, and the third-most-significant bit is `0` to indicate that it is not an MPT. The remaining 61 bits represent the quantity of XRP. Since the maximum amount of XRP (10<sup>17</sup> drops) only requires 57 bits, you can calculate XRP serialized format by taking standard 64-bit unsigned integer and performing a bitwise-OR with `0x4000000000000000`.

- **Fungible Tokens**

    Fungible tokens (also called "IOUs") consist of three segments in order:

>     1. 64 bits indicating the amount in the [token amount format](#token-amount-format). The first bit is `1` to indicate that this is not XRP nor an MPT.

Reviewer: ```suggestion
    1. 64 bits indicating the amount in the [token amount format](#token-amount-format). The first bit is `1` to indicate that this is an IOU.
```
Is there a reason we have to say this in the negative of what it's not? Like there may be a future scenarior where `1` can represent an IOU and something else, while still not being an MPT or XRP?

- **MPTs**

    Multi-Purpose Tokens (MPTs) consist of four segments in order:

    1. 8 bits indicating that this is an MPT. The most significant bit is `0` to indicate that it's not a fungible token. The second bit is `1` to indicate that it is postiive. The third most significant bit is `1` to indicate that it is an MPT. The remaining 5 bits are reserved and must all be `0`. In other words, the first byte is `0x60`.
    2. 64 bits indicating the quantity of the MPT, as a 64-bit unsigned integer. (However, the maximum amount cannot be larger than 2<sup>63</sup>-1.)
    3. 32 bits indicating the `Sequence` number of the transaction that created the MPT issuance.
    4. 160 bits indicating the [AccountID][] of the MPT's issuer.

You can tell which of the three sub-types an amount is based on the first and third most significant bits: 

    - If the first bit is a `1`, it's a fungible token (IOU).
    - If the first bit and third bit are both `0`, it's XRP.
    - If the first bit is a `0` and the third bit is a `1`, it's an MPT.

{% admonition type="warning" title="Caution" %}

> Not all types of amount are valid in all places. Some fields can only represent XRP, or XRP and fungible tokens but not MPTs. These limitations are defined by the individual transactions and ledger entries.

Reviewer: ```suggestion
Not all `STAmount` types are valid in all places. Some fields can only represent XRP, or XRP and fungible tokens but not MPTs. These limitations are defined by the individual transactions and ledger entries.
```
I think writing it this way makes it easier to understand we're talking about defining amount fields. Admittedly very minor, but I could see how someone just learning about this might stumble momentarily.

Author: I am trying to get away from using the `ST` prefix in docs, and I don't see the clarification you're trying to make here.

You can't use MPTs in Escrow, for example. And you can't use fungible tokens until the TokenEscrow amendment gets enabled, either. Similarly, you can't deposit MPTs into AMMs. Those are the types of limitations I'm trying to mention here.

{% /admonition %}

> The following diagram shows the serialization formats for both XRP amounts and token amounts: ***TODO: update***

Reviewer: ```suggestion
The following diagram shows the serialization formats for both XRP amounts and token amounts:***
```
Does this diagram need to be updated still?

Author: It did. Done now.

The **nonstandard format** is any 160 bits of data as long as the first 8 bits are not `0x00`.

#### MPT Amount Format

> The following diagram shows the amount serialization for MPTs. ***TODO: consolidate or not?***

Reviewer: ```suggestion
The following diagram shows the amount serialization for MPTs.
```
Missed note?

Author: Removed this redundant section entirely.

### Blob Fields
[Blob]: #blob-fields

The _Blob_ type (also called "STBlob") is a [length-prefixed](#length-prefixing) field with arbitrary data. Two common fields that use this type are `SigningPubKey` and `TxnSignature`, which contain (respectively) the public key and signature that authorize a transaction to be executed.

Blob fields have no further structure to their contents, so they consist of exactly the amount of bytes indicated in the variable-length encoding, after the Field ID and length prefixes.

### Currency Fields
[Currency]: #currency-fields

> Some fields specify a currency code, which could be an a fungible token, the ticker symbol for an off-ledger asset, or some other identifier for a currency. This field type is currently used only in [Price Oracles](../../concepts/decentralized-storage/price-oracles.md).

Reviewer: ```suggestion
Some fields specify a currency code, which could be a fungible token, the ticker symbol for an off-ledger asset, or some other identifier for a currency. This field type is currently used only in [Price Oracles](../../concepts/decentralized-storage/price-oracles.md).
```

> These fields consists of 160 bits of binary data. If the data matches the ["standard" currency code format](#currency-codes), it may be represented as a three-letter currency code string in JSON. Otherwise, it is represented as hexadecimal. Client libraries _may_ attempt to interpret this as a string of ASCII or UTF-8, but it is not guaranteed to be valid.

Reviewer: ```suggestion
These fields consist of 160 bits of binary data. If the data matches the ["standard" currency code format](#currency-codes), it may be represented as a three-letter currency code string in JSON. Otherwise, it is represented as hexadecimal. Client libraries _may_ attempt to interpret this as a string of ASCII or UTF-8, but it is not guaranteed to be valid.
```
Do we not have a definitive answer here for which client libraries interpret as ASCII or UTF-8? Since odd to leave it so ambiguous.

Author: It's up to different client libraries. Some of them do, some don't, and that may change. I don't have a survey here. But I can link the code sample we have for this purpose.
